EmailLogCleanup/src/main/java/org/vm/shared/S3LogWriter.java

```diff
@@ -0,0 +1,50 @@
+package org.vm.shared;
+
+import com.amazonaws.SdkClientException;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PutObjectRequest;
+
+import java.io.InputStream;
+
+/**
+ * The S3LogWriter class takes is a class that implements the WriteLog interface and specifiying it
+ * to write to a file to an s3 bucket.
+ */
+public class S3LogWriter implements WriteLog{
+  
+  private final String bucket;
+  
+  /**
+   * The constructor takes in the name of the destination S3 bucket.
+   * @param bucketName is the name of the destination bucket in AWS.
+   */
+  public S3LogWriter(String bucketName) {
+    bucket = bucketName;
+  }
+  
+  /**
+   * Method inherited from the WriteLog interface that will create a put request for a file
+   * into an s3 bucket.
+   */
+  @Override
+  public void writeLog(InputStream stream, String filename) {
+    //Sets up the s3 bucket.
+    AmazonS3 s3Client = AmazonS3ClientBuilder.defaultClient();
+  
+    //Creates the PutObject request.
+    try {
+      PutObjectRequest request
+          = new PutObjectRequest(bucket, filename, stream, new ObjectMetadata());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: This code is uploading an object of type stream to S3 without setting the content length of the object. Failure to specify a content length will cause the entire contents of the input stream to be buffered locally in memory so that the content length can be calculated, which can result in negative performance problems.

**Fix**: Specify the content length for the data stream in the object metadata parameter.

[Learn more](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/PutObjectRequest.html#PutObjectRequest-java.lang.String-java.lang.String-java.io.InputStream-com.amazonaws.services.s3.model.ObjectMetadata-)